File: Algebra_Tests.rb
require 'test/unit'
require 'SLL'
require 'Algebra'
require 'SLLParser'

class Algebra_Tests < Test::Unit::TestCase
  include SLL
  include SLLParser
  include Algebra
  def to_s_test(expected, e)
    assert_equal(expected, e.to_s)
  end

  def test001_to_s()
    to_s_test('x', Var.new("x"))
    to_s_test('A(x,y)', Ctr.new("A", [Var.new("x"), Var.new("y")]))
    to_s_test('fX(x,y)', FCall.new("fX", [Var.new("x"), Var.new("y")]))
    to_s_test('gX(x,y)', GCall.new("gX", [Var.new("x"), Var.new("y")]))
    to_s_test('let x=y in y', Let.new(Var.new("y"), [["x", Var.new("y")]]))
  end

  def matchOK(pat, exp, expected)
    subst = matchAgainst(pExp(pat), pExp(exp))
    if subst != nil
      subst = subst.sort.map { |kv| "#{kv[0]}->#{kv[1]};"} * ''
    end
    assert_equal(expected, subst)
  end

  def matchNo(pat, exp)
    subst = matchAgainst(pExp(pat), pExp(exp))
    assert_equal(nil, subst)
  end

  def test101MatchV_E()
    matchOK("x", "S(Z)", "x->S(Z);")
  end

  def test102MatchC_V()
    matchNo("Z", "x")
  end

  def test103MatchC_C()
    matchOK("C(x,y)", "C(A,B)", "x->A;y->B;")
  end

  def test104MatchC1_C2()
      matchNo("C(x,y)", "D(A,B)")
  end
  
  def test105MatchC_F()
      matchNo("C(x,y)", "f(A,B)")
  end
  
  def test106MatchX_X_Eq()
      matchOK("C(x,x)", "C(A,A)", "x->A;")
  end
  
  def test107Match_X_XY()
      matchNo("C(x,y)", "C(A,B,C)")
  end
  
  def testMatch_XY_X()
      matchNo("C(x,y,z)", "C(A,B)")
  end

  def equivYes(e1, e2)
      assert(equiv(pExp(e1), pExp(e2)))
  end
  
  def test201EquivYes()
      equivYes("gA(fB(x,y),C)", "gA(fB(a,b),C)")
  end
  
  def equivNo(e1, e2)
      assert(!equiv(pExp(e1), pExp(e2)))
  end
  
  def test301EquivNo()
      equivNo("gA(fB(x,y),x)", "gA(fB(a,a),b)")
  end

end

File: src/spsc/tests/Algebra_Tests.scala
package spsc.tests

import org.junit.Test
import org.junit.Assert._
import Algebra._

class Algebra_Tests {

  @Test def test101TheSameFunctor() : Unit = {
    assertTrue(shellEq(Ctr("A", List()), Ctr("A", List())))
    assertTrue(shellEq(FCall("A", List()), FCall("A", List())))
    assertTrue(shellEq(GCall("A", List()), GCall("A", List())))
    assertFalse(shellEq(Ctr("A", List()), Ctr("B", List())))
    assertFalse(shellEq(Ctr("A", List()), FCall("A", List())))
    assertFalse(shellEq(Ctr("A", List()), Ctr("A", List(Var("y")))))
  }
  
  @Test def test201Subst() : Unit = {
    val e1 = SParsers.parseTerm("E1()")
    val e2 = SParsers.parseTerm("E2()")
    val e = SParsers.parseTerm("Cons(x1,Cons(x2,Cons(x3,Nil())))")
    val subst = Map(Var("x1")->e1, Var("x2")->e2)
    assertEquals("Cons(E1(),Cons(E2(),Cons(x3,Nil())))",
                 applySubst(subst, e).toString())
  }

  @Test def test302Vars() : Unit = {
    val e = SParsers.parseTerm("A(x,B(y,z),a)")
    assertEquals(List(Var("x"), Var("y"), Var("z"), Var("a")), vars(e))
    
    val e1 = SParsers.parseTerm("A(x,B(y,x),a)")
    assertEquals(List(Var("x"), Var("y"), Var("a")), vars(e1))
  }

  def substToString(subst : Map[Var, Term]) : String = {
    if( subst == null)
      null
    else {
      var acc = ""
      for((v, e) <- subst)
        acc += v.toString() + "->" + e.toString() + ";"
        acc.mkString("")
    }
  }
  
  def matchOK(pat : String, exp : String, expected : String) : Unit = {
    val subst = matchAgainst(SParsers.parseTerm(pat), SParsers.parseTerm(exp))
    assertEquals(expected, substToString(subst))
  }

  def matchNo(pat : String, exp : String) : Unit = {
    val subst = matchAgainst(SParsers.parseTerm(pat), SParsers.parseTerm(exp))
    assertEquals(null, substToString(subst))
  }

  @Test def test401MatchV_E() : Unit = {
    matchOK("x", "S(Z())", "x->S(Z());")
  }

  @Test def test402MatchC_V() : Unit = {
    matchNo("Z()", "x")
  }

  @Test def test403MatchC_C() : Unit = {
    matchOK("C(x,y)", "C(A(),B())", "x->A();y->B();")
  }

  @Test def test404MatchC1_C2() : Unit = {
    matchNo("C(x,y)", "D(A(),B())")
  }

  @Test def test405MatchC_F() : Unit = {
    matchNo("C(x,y)", "f(A,B)")
  }

  @Test def test406MatchX_X_Eq() : Unit = {
    matchOK("C(x,x)", "C(A(),A())", "x->A();")
  }

  @Test def test407Match_X_XY() : Unit = {
    matchNo("C(x,y)", "C(A(),B(),C())")
  }

  @Test def testMatch_XY_X() : Unit = {
    matchNo("C(x,y,z)", "C(A(),B())")
  }

  def equivYes(e1 : String, e2 : String) : Unit = {
    assertTrue(equiv(SParsers.parseTerm(e1), SParsers.parseTerm(e2)))
  }

  @Test def test501EquivYes() : Unit = {
    equivYes("gA(fB(x,y),C)", "gA(fB(a,b),C)")
  }

  def equivNo(e1 : String, e2 : String) : Unit = {
    assertFalse(equiv(SParsers.parseTerm(e1), SParsers.parseTerm(e2)))
  }

  @Test def test502EquivNo() : Unit = {
    equivNo("gA(fB(x,y),x)", "gA(fB(a,a),b)")
  }
}
